File: WindowsApp3/Classes/DataOperations.vb
Imports System.Data.SqlClient
Imports System.IO
Imports BaseConnectionLibrary.ConnectionClasses

Namespace Classes

    Public Class DataOperations
        Inherits SqlServerConnection

        Private _excelCompanyFileName As String = Path.Combine(
            AppDomain.CurrentDomain.BaseDirectory, "Customers.xlsx")

        Public Function CopyToApplicationFolder(pFileName As String) As Boolean

            Try

                File.Copy(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, $"Files\{Path.GetFileName(pFileName)}"),
                          Path.Combine(AppDomain.CurrentDomain.BaseDirectory, pFileName), True)

                Return True

            Catch e1 As Exception
                Return False
            End Try

        End Function
        Public Sub WriteCustomerTableToXmlFile(pFileName As String, ByRef pRowsExported As Integer)

            CopyToApplicationFolder(_excelCompanyFileName)

            ' server name
            DatabaseServer = "KARENS-PC"
            ' database name
            DefaultCatalog = "ExcelExporting"

            Using cn = New SqlConnection With {.ConnectionString = ConnectionString}
                Using cmd = New SqlCommand With {.Connection = cn}


                    cmd.CommandText = "INSERT INTO OPENROWSET('Microsoft.ACE.OLEDB.12.0', " &
                                      $"'Excel 12.0;Database={pFileName}','SELECT * from [Customers$]') " &
                                      "SELECT CustomerIdentifier,CompanyName,ContactName,ContactTitle," &
                                      "[Address],City,Region,PostalCode,Country,Phone " &
                                      "FROM Customers"

                    Try
                        cn.Open()
                        pRowsExported = cmd.ExecuteNonQuery()

                    Catch ex As Exception
                        mHasException = True
                        mLastException = ex
                    End Try
                End Using
            End Using
        End Sub
    End Class
End Namespace

File: WindowsApp3a/Classes/DataOperations.cs
using System;
using System.Data.SqlClient;
using System.IO;
using BaseConnectionLibrary.ConnectionClasses;

namespace WindowsApp3a.Classes
{
    /// <summary>
    /// https://code.msdn.microsoft.com/SQL-Server-export-to-Excel-c3c6f21f
    /// </summary>
    public class DataOperations : SqlServerConnection
    {
        public DataOperations()
        {
            DefaultCatalog = "ExcelExporting";
        }
        /// <summary>
        /// Copy template excel file to app folder
        /// </summary>
        /// <param name="pFileName"></param>
        /// <returns></returns>
        public bool CopyToApplicationFolder(string pFileName)
        {

            try
            {
                File.Copy(
                    Path.Combine(AppDomain.CurrentDomain.BaseDirectory, $"Files\\{Path.GetFileName(pFileName)}"), 
                    Path.Combine(AppDomain.CurrentDomain.BaseDirectory, pFileName), true);

                return true;

            }
            catch (Exception)
            {
                return false;
            }

        }
        /// <summary>
        /// Copy template file to app folder with a different name
        /// </summary>
        /// <param name="pFileName"></param>
        /// <param name="pTargetFile"></param>
        /// <returns></returns>
        public bool CopyToApplicationFolder(string pFileName, string pTargetFile)
        {

            try
            {
                File.Copy(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, $"Files\\{Path.GetFileName(pFileName)}"), 
                    Path.Combine(AppDomain.CurrentDomain.BaseDirectory, pTargetFile), true);

                return true;

            }
            catch (Exception e)
            {
                mHasException = true;
                mLastException = e;
            }

            return IsSuccessFul;
        }
        /// <summary>
        /// Export data to Excel
        /// </summary>
        /// <param name="pFileName"></param>
        /// <param name="pRowsExported">Used to return export row count </param>
        /// <returns></returns>
        public bool WriteCustomerTableToXmlFile(string pFileName, ref int pRowsExported)
        {
            using (var cn = new SqlConnection { ConnectionString = ConnectionString })
            {
                using (var cmd = new SqlCommand { Connection = cn })
                {

                    cmd.CommandText = 
                        $"INSERT INTO OPENROWSET('Microsoft.ACE.OLEDB.12.0', 'Excel 12.0;Database={pFileName}'," + 
                        "'SELECT * from [Customers$]') " + 
                        "SELECT CustomerIdentifier,CompanyName,ContactName,ContactTitle," + 
                        "[Address],City,Region,PostalCode,Country,Phone " + 
                        "FROM Customers";

                    Console.WriteLine(cmd.CommandText);
                    cn.Open();

                    try
                    {

                        pRowsExported = cmd.ExecuteNonQuery();
                        return pRowsExported > 0;

                    }
                    catch (Exception e)
                    {
                        mHasException = true;
                        mLastException = e;
                    }
                }
            }

            return IsSuccessFul;
        }
    }
}
